vpr/src/place/initial_placement.cpp

```diff
@@ -549,47 +547,188 @@ static std::vector<ClusterBlockId> find_centroid_loc(const t_pl_macro& pl_macro,
+        // In order to ensure our BFS finds the closest compatible location, we
+        // traverse compressed grid locations which may not actually be valid
+        // (i.e. no tile exists there). This is fine, we just need to check for
+        // them to ensure we never try to put a cluster there.
+        bool is_valid_compressed_loc = false;
+        const auto& compressed_col_blk_map = compressed_block_grid.get_column_block_map(loc.x, 0);
+        if (compressed_col_blk_map.count(loc.y) != 0)
+            is_valid_compressed_loc = true;
+
+        // If this distance is better than the best we have seen so far, try
+        // to see if this is a better solution.
+        if (is_valid_compressed_loc) {
+            // Check if a sub-tile is available at this location.
+            const t_physical_tile_type_ptr phy_type = g_vpr_ctx.device().grid.get_physical_type(grid_loc);
+            const auto& compatible_sub_tiles = compressed_block_grid.compatible_sub_tile_num(phy_type->index);
+            int new_sub_tile = -1;
+            for (int sub_tile : compatible_sub_tiles) {
+                if (blk_loc_registry.grid_blocks().is_sub_tile_empty(grid_loc, sub_tile)) {
+                    new_sub_tile = sub_tile;
+                    break;
+                }
+            }
+            if (new_sub_tile != -1) {
+                // If a sub-tile is available, set this to be the first sub-tile
+                // available and check if this site is legal for this macro.
+                // TODO: Should we pick a random sub-tile instead?
+                // Note: We are usin the fully legality check here to check for
+                //       floorplanning constraints and compatibility for all
+                //       members of the macro. This prevents some macros being
+                //       placed where they obviously cannot be implemented.
+                t_pl_loc new_loc = t_pl_loc(grid_loc.x, grid_loc.y, new_sub_tile, grid_loc.layer_num);
+                bool site_legal_for_macro = macro_can_be_placed(pl_macro,
+                                                                new_loc,
+                                                                true /*check_all_legality*/,
+                                                                blk_loc_registry);
+                if (site_legal_for_macro) {
+                    // Update the best solition.
+                    // Note: We need to keep searching since the compressed grid
+                    //       may present a location which is closer in compressed
+                    //       space earlier than a location which is closer in
+                    //       grid space.
+                    best_dist = grid_dist;
+                    best_loc = new_loc;
+                }
+            }
+        }
+
+        // Push the neighbors (in the compressed grid) onto the queue.
+        if (loc.x > 0) {
+            t_physical_tile_loc new_loc = t_physical_tile_loc(loc.x - 1,
```

Comment:
I'd work comp_ or compressed_ into the name of this variable (e.g. new_comp_loc or new_compressed_loc) as it is easy to get confused as to what grid we're looking things up in.